server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -835,432 +835,550 @@ public void deleteSnapshots(
-     * @param blobsToDelete Blob names in the shard directory that have become unreferenced in the new shard generation
+     * <p>
+     *     Represents the process of deleting some collection of snapshots within this repository which since 7.6.0 looks like this:
+     * </p>
+     * <ul>
+     *     <li>Write a new {@link BlobStoreIndexShardSnapshots} for each affected shard, and compute the blobs to delete.</li>
+     *     <li>Update the {@link RepositoryData} to remove references to deleted snapshots/indices and point to the new
+     *     {@link BlobStoreIndexShardSnapshots} files.</li>
+     *     <li>Remove up any now-unreferenced blobs.</li>
+     * </ul>
+     * <p>
+     *     Until the {@link RepositoryData} is updated there should be no other activities in the repository, and in particular the root
+     *     blob must not change until it is updated by this deletion and {@link SnapshotDeleteListener#onRepositoryDataWritten} is called.
+     * </p>
      */
-    private record ShardSnapshotMetaDeleteResult(
-        IndexId indexId,
-        int shardId,
-        ShardGeneration newGeneration,
-        Collection<String> blobsToDelete
-    ) {}
+    class SnapshotsDeletion {
+
+        /**
+         * The IDs of the snapshots to delete.
+         */
+        private final Collection<SnapshotId> snapshotIds;
+
+        /**
+         * The {@link RepositoryData} generation at the start of the process, to ensure that the {@link RepositoryData} does not change
+         * while the new {@link BlobStoreIndexShardSnapshots} are being written.
+         */
+        private final long originalRepositoryDataGeneration;
+
+        /**
+         * The minimum {@link IndexVersion} of the nodes in the cluster and the snapshots remaining in the repository. The repository must
+         * remain readable by all node versions which support this {@link IndexVersion}.
+         */
+        private final IndexVersion repositoryFormatIndexVersion;
+
+        /**
+         * Whether the {@link #repositoryFormatIndexVersion} is new enough to support naming {@link BlobStoreIndexShardSnapshots} blobs with
+         * UUIDs (i.e. does not need to remain compatible with versions before v7.6.0). Older repositories use (unsafe) numeric indices for
+         * these blobs instead.
+         */
+        private final boolean useShardGenerations;
+
+        /**
+         * All blobs in the repository root at the start of the operation, obtained by listing the repository contents. Note that this may
+         * include some blobs which are no longer referenced by the current {@link RepositoryData}, but which have not yet been removed by
+         * the cleanup that follows an earlier deletion. This cleanup may or may not still be ongoing (it could have been running on a
+         * different node, which died before completing it) so we track all the blobs here and clean them up again at the end.
```

Comment:
A question: By "still be onging", does it mean the async part of the snapshot file deletion could be still "running"? The example says the node could have died before completing. But is it possible that the node is still running the async part or does it stop the async part when it steps down from being master? Is it also possible that a previous async deletion is still running by the "current' node?

Reply:
Yep all are possible. There is no way we can prevent another node from becoming master while we are in the process of deleting a blob, so we don't even try. I expanded the wording of the comment a bit to clarify, see c8c221b1a05b1a4a30cdb9f7146c8a287c5fba8f.